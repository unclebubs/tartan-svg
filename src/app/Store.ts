import { type EnhancedStore, combineReducers, configureStore } from '@reduxjs/toolkit'
import counterReducer from '../model/tartan/TartanSlice'

// Create the root reducer separately so we can extract the RootState type
const rootReducer = combineReducers({
  counter: counterReducer
})

const store = configureStore({
  reducer: {
    counter: counterReducer
  }
})

export const setupStore = (preloadedState?: Partial<RootState>): EnhancedStore => {
  return configureStore({
    reducer: rootReducer,
    preloadedState
  })
}

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = typeof store.dispatch

export default store
