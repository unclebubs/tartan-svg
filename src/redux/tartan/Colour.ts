export class Colour {
  colourCode: string
  hex: string
  name: string

  constructor (colourCode: string, hex: string, name: string) {
    this.colourCode = colourCode
    this.hex = hex
    this.name = name
  }
}
