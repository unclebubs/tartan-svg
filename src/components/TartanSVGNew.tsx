import React, { type ReactElement } from 'react'
import { Weft } from './Weft'
import { Warp } from './Warp'
import { useAppSelector } from '../app/hooks'

interface TartanProps {
  style?: React.CSSProperties
  id?: string
  useBlur?: boolean
  blurValue?: number
}

const TartanSVGNew: React.FC<TartanProps> = (props) => {
  const { style, id, useBlur = false, blurValue = 1 } = props
  const { tartan } = useAppSelector((state) => state.counter)

  const filter = (useBlur: boolean): ReactElement | null => {
    if (useBlur) {
      return (
        <filter id="blur">
          <feGaussianBlur stdDeviation={blurValue} in="SourceGraphic" result="BLUR"></feGaussianBlur>
        </filter>
      )
    }
    return null
  }

  return (
    <svg id={id} height='1000' width='1000' style={style} xmlns="http://www.w3.org/2000/svg">
      { filter(useBlur) }
      <defs>
        <mask id="threadHatchingMask" x="0" y="0" width="1" height="1">
          <rect x="0" y="0" width="100%" height="100%" fill="url(#threadHatching)"></rect>
        </mask>
        <pattern id="threadHatching" x="0" y="0" patternUnits="userSpaceOnUse" width="4" height="4">
          <rect x="0" y="-1" height="2" width="1" fill="white"></rect>
          <rect x="0" y="3" height="2" width="1" fill="white"></rect>
          <rect x="1" y="0" height="2" width="1" fill="white"></rect>
          <rect x="2" y="1" height="2" width="1" fill="white"></rect>
          <rect x="3" y="2" height="2" width="1" fill="white"></rect>
        </pattern>
        <pattern id="sett" x="0" y="0" patternUnits="userSpaceOnUse" width={tartan?.getSetSize()} height={tartan?.getSetSize()} >
          <Weft />
          <Warp />
        </pattern>
      </defs>
      <rect x="0" y="0" height="100%" width="100%" fill="url(#sett)" filter={ useBlur ? 'url(#blur)' : ''}></rect>
    </svg>
  )
}
export default TartanSVGNew
