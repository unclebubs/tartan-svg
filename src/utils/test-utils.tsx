import React, { type PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import { Provider } from 'react-redux'

import { setupStore } from '../app/Store'
import type { AppStore, RootState } from '../app/Store'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>
  store?: AppStore
}

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type, @typescript-eslint/space-before-function-paren
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  // eslint-disable-next-line @typescript-eslint/space-before-function-paren, @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}
