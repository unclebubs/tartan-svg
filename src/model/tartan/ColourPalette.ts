import { Colour } from './Colour'

export class ColourPalette {
  paletteStr: string
  colours: Map<string, Colour >
  constructor (paletteStr: string) {
    this.paletteStr = paletteStr
    this.colours = ColourPalette.buildPalette(this.paletteStr)
  }

  getColour (colourCode: string): Colour | undefined {
    return this.colours.get(colourCode)
  }

  static buildPalette (paletteStr: string): Map<string, Colour> {
    const colourMap = new Map<string, Colour>()
    const palette = paletteStr.slice(0, paletteStr.length - 1) // knock off last ":"
    const colors = palette.split(';')
    colors.forEach(color => {
      const equalsIndex = color.indexOf('=')
      const colorCode = color.substring(0, equalsIndex).trim()
      const hex = color.substring(equalsIndex + 1, equalsIndex + 7).trim()
      const name = color.substring(equalsIndex + 7).trim()
      const colour = new Colour(colorCode, hex, name)
      colourMap.set(colour.colourCode, colour)
    })

    return colourMap
  }
}
