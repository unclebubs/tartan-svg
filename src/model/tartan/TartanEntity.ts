import { ColourPalette } from './ColourPalette'
import { Thread } from './Thread'

export class TartanEntity {
  static halfSettToken: string = '/'

  name: string
  readonly threads: Thread[] = []
  readonly fullSet: Thread[] = []
  readonly colourPalette: ColourPalette | null = null
  readonly isHalfSet: boolean = false
  readonly noOfThreads: number = 0
  readonly scaleFactor: number = 1
  readonly scaledStrokeLength: number = 1
  strokeLength: number = 2
  noOfSetts: number
  imageSize: number
  _threadCount: string
  _palette: string
  xOffsetThreadCount: number
  yOffsetThreadCount: number
  useBlur: boolean
  blurValue: number

  constructor (
    name: string,
    threadCount: string,
    palette: string,
    noOfSetts: number,
    imageSize: number,
    xOffsetThreadCount: number,
    yOffsetThreadCount: number,
    useBlur: boolean,
    blurValue: number = 0
  ) {
    this.name = name
    this._threadCount = threadCount
    this._palette = palette
    this.noOfSetts = noOfSetts
    this.imageSize = imageSize
    this.xOffsetThreadCount = xOffsetThreadCount
    this.yOffsetThreadCount = yOffsetThreadCount
    this.useBlur = useBlur
    this.blurValue = blurValue
    if (name.length > 0 && threadCount.length > 0 && palette.length > 0) {
      this.colourPalette = new ColourPalette(this._palette)
      this.threads = TartanEntity.extractThreadsFromThreadCount(this._threadCount, this.colourPalette)
      this.isHalfSet = this.updateIsHalfSet()
      this.fullSet = this.buildFullSet(this.threads, this.isHalfSet)
      this.noOfThreads = this.updateNumberOfThreads()
      this.scaleFactor = (this.imageSize / (this.noOfSetts * this.getSetSize()))
      this.scaledStrokeLength = this.scaleFactor * this.strokeLength
    }
  }

  static addSpaceAfterNumbers (str: string): string {
    return str.replace(/(\d+)/g, '$1 ').replace(/ {2,}/g, ' ').trim()
  }

  buildFullSet = (threads: Thread[], isHalfSet: boolean): Thread[] => {
    let fSet: Thread[] = []
    if (isHalfSet) {
      const nextSet = threads.toReversed().slice(1, threads.length - 1)
      fSet = threads.concat(nextSet)
    } else {
      fSet = [...threads]
    }

    return fSet
  }

  updateIsHalfSet = (): boolean => {
    return this.threads[this.threads.length - 1].isPivot
  }

  updateNumberOfThreads = (): number => {
    return this.fullSet.reduce(
      (prev, thread) => prev + thread.noOfThreads, 0
    )
  }

  getSetSize (): number {
    return this.strokeLength * this.noOfThreads
  }

  static extractThreadsFromThreadCount (threadCount: string, colourPalette: ColourPalette): Thread[] {
    const threadCountItems: Thread [] = []
    const threadCountTokens: string[] = this.addSpaceAfterNumbers(threadCount.toUpperCase()).trim().split(' ')
    threadCountTokens.forEach(threadItem => {
      const newThreadItem = threadItem.replace(/(\d+)/g, (_, num) => {
        return '=' + num
      })

      const threadParts = newThreadItem.trim().split('=')
      if (threadParts[0].length > 0) {
        let colourCode = threadParts[0]
        const isPivot = colourCode.includes(TartanEntity.halfSettToken)
        colourCode = isPivot ? colourCode.substring(0, colourCode.indexOf('/')) : colourCode
        const noOfThreads = parseInt(threadParts[1])
        const colour = colourPalette.getColour(colourCode)
        threadCountItems.push(new Thread('dummy', colour, colourCode, noOfThreads, isPivot))
      }
    })
    return threadCountItems
  }

  set palette (palette: string) {
    this._palette = palette
  }

  get palette (): string {
    return this._palette
  }

  set threadCount (val: string) {
    this._threadCount = val
  }

  get threadCount (): string {
    return this._threadCount
  }
}
