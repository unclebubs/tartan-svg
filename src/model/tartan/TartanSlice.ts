import { createSlice, type PayloadAction } from '@reduxjs/toolkit'
import { TartanEntity } from './TartanEntity'

export interface UpdateTartanActionType {
  name: string
  threadCount: string
  colourPalette: string
  noOfSetts: number
  imageSize: number
  xOffsetThreadCount: number
  yOffsetThreadCount: number
}

export interface UpdateOffsetActionType {
  xOffsetThreadCount: number
  yOffsetThreadCount: number
}

export interface UpdateFilterActionType {
  isUseBlurFilter: boolean
  blurValue: number
}

interface TartanState {
  name: string
  threadCount: string
  colourPalette: string
  imageSize: number
  noOfSetts: number
  tartan: TartanEntity
  isUseBlurFilter: boolean
  blurValue: number
}

const initialState: TartanState = {
  name: '',
  threadCount: '',
  colourPalette: '',
  imageSize: 0,
  noOfSetts: 1,
  tartan: new TartanEntity('', '', '', 0, 0, 0, 0),
  isUseBlurFilter: true,
  blurValue: 0.5
}

export const counterSlice = createSlice({
  name: 'tartan',
  initialState,
  reducers: {
    update: (state, action: PayloadAction<UpdateTartanActionType>) => {
      const { name, threadCount, colourPalette, noOfSetts, imageSize, xOffsetThreadCount, yOffsetThreadCount } = action.payload
      state.tartan = new TartanEntity(name, threadCount, colourPalette, noOfSetts, imageSize, xOffsetThreadCount, yOffsetThreadCount)
    },
    updateOffset: (state, action: PayloadAction<UpdateOffsetActionType>) => {
      const { xOffsetThreadCount, yOffsetThreadCount } = action.payload
      const { name, threadCount, palette, noOfSetts, imageSize } = state.tartan
      const newT = new TartanEntity(name, threadCount, palette, noOfSetts, imageSize, xOffsetThreadCount, yOffsetThreadCount)
      state.tartan = newT
    },
    updateFilter: (state, action: PayloadAction<UpdateFilterActionType>) => {
      const { isUseBlurFilter, blurValue } = action.payload
      state.isUseBlurFilter = isUseBlurFilter
      state.blurValue = blurValue
    }

  }
})

// Action creators are generated for each case reducer function
export const { update, updateOffset, updateFilter } = counterSlice.actions

export default counterSlice.reducer
