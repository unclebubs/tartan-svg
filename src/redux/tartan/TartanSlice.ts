import { createSlice, type PayloadAction } from '@reduxjs/toolkit'
import { TartanEntity } from './TartanEntity'

export interface UpdateTartanActionType {
  name: string
  threadCount: string
  colourPalette: string
  noOfSetts: number
  imageSize: number
}

interface TartanState {
  name: string
  threadCount: string
  colourPalette: string
  imageSize: number
  noOfSetts: number
  tartan: TartanEntity | null
}

const initialState: TartanState = {
  name: 'Abercrombie Ancient',
  threadCount: 'G56 W12 B/28',
  colourPalette: 'G=006818GREEN;K=101010BLACK;CW=FCFCFCCLEAR;DB=202060DARK BLUE;R=C80000RED;',
  imageSize: 500,
  noOfSetts: 1,
  tartan: null
}

export const counterSlice = createSlice({
  name: 'tartan',
  initialState,
  reducers: {
    update: (state, action: PayloadAction<UpdateTartanActionType>) => {
      const { name, threadCount, colourPalette, noOfSetts, imageSize } = action.payload
      console.log('The payload is ', action.payload)
      state.tartan = new TartanEntity(name, threadCount, colourPalette, noOfSetts, imageSize)
    }
  }
})

// Action creators are generated for each case reducer function
export const { update } = counterSlice.actions

export default counterSlice.reducer
