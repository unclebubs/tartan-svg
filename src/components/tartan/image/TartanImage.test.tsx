import React from 'react'
import { setupStore } from '../../../app/Store'
import { update, updateFilter, type UpdateTartanActionType } from '../../../redux/tartan/TartanSlice'
import { renderWithProviders } from '../../../utils/test-utils'
import TartanSVGNew from './TartanImage'

const TEST_TARTAN_PARAMS: UpdateTartanActionType = {
  name: 'Test Tartan',
  threadCount: 'G56 W12 B/28',
  colourPalette: 'G=006818GREEN;W=FCFCFCCLEAR;B=202060DARK BLUE;',
  noOfSetts: 1,
  imageSize: 1000,
  xOffsetThreadCount: 0,
  yOffsetThreadCount: 0
}

test('renders correctly', async () => {
  const store = setupStore()
  store.dispatch(
    update(TEST_TARTAN_PARAMS)
  )
  store.dispatch(updateFilter({
    isUseBlurFilter: false,
    blurValue: 0
  }))

  const tartanSVG = renderWithProviders(<TartanSVGNew id='testId' style={{ backgroundColor: 'red' }} />, { store })
  expect(tartanSVG).toMatchSnapshot()
})

test('does not render or reference filter when isUseBlurFilter is false', async () => {
  const store = setupStore()
  store.dispatch(update(TEST_TARTAN_PARAMS))
  store.dispatch(updateFilter({
    isUseBlurFilter: false,
    blurValue: 0
  }))

  const tartanSVG = renderWithProviders(<TartanSVGNew />, { store })
  expect(tartanSVG.queryByTestId('blurFilter')).toBeNull()
  expect(tartanSVG.queryByTestId('tartanRect')?.getAttribute('filter')).toBe('')
})

test('x and y offsets are applied correctly when imageSize matches settSize', async () => {
  const store = setupStore()
  // set image size to match sett size so offsets will just equal actual values
  store.dispatch(update({ ...TEST_TARTAN_PARAMS, xOffsetThreadCount: 11, yOffsetThreadCount: 3, imageSize: 108 }))
  const tartanSVG = renderWithProviders(<TartanSVGNew />, { store })
  expect(tartanSVG.getByTestId('testSett').getAttribute('x')).toBe('11')
  expect(tartanSVG.getByTestId('testSett').getAttribute('y')).toBe('3')
})

test('x and y offsets are applied correctly when imageSize doesnt match settSize', async () => {
  const store = setupStore()
  // set image size to match 2 times sett size so offsets will just equal double values
  store.dispatch(update({ ...TEST_TARTAN_PARAMS, xOffsetThreadCount: 11, yOffsetThreadCount: 3, imageSize: 216 }))
  const tartanSVGb = renderWithProviders(<TartanSVGNew />, { store })
  expect(tartanSVGb.getByTestId('testSett').getAttribute('x')).toBe('22')
  expect(tartanSVGb.getByTestId('testSett').getAttribute('y')).toBe('6')
})
