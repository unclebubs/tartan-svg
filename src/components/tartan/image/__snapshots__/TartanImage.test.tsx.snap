// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TartanImage renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <svg
        data-testid="tartanImage"
        height="1000"
        id="testId"
        style="background-color: red;"
        width="1000"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <mask
            height="1"
            id="threadHatchingMask"
            width="1"
            x="0"
            y="0"
          >
            <rect
              fill="url(#threadHatching)"
              height="100%"
              width="100%"
              x="0"
              y="0"
            />
          </mask>
          <pattern
            height="18.51851851851852"
            id="threadHatching"
            patternUnits="userSpaceOnUse"
            width="18.51851851851852"
            x="0"
            y="0"
          >
            <rect
              fill="white"
              height="9.25925925925926"
              width="4.62962962962963"
              x="0"
              y="-4.62962962962963"
            />
            <rect
              fill="white"
              height="9.25925925925926"
              width="4.62962962962963"
              x="0"
              y="13.88888888888889"
            />
            <rect
              fill="white"
              height="9.25925925925926"
              width="4.62962962962963"
              x="4.62962962962963"
              y="0"
            />
            <rect
              fill="white"
              height="9.25925925925926"
              width="4.62962962962963"
              x="9.25925925925926"
              y="4.62962962962963"
            />
            <rect
              fill="white"
              height="9.25925925925926"
              width="4.62962962962963"
              x="13.88888888888889"
              y="9.25925925925926"
            />
          </pattern>
          <pattern
            data-testid="testSett"
            height="1000"
            id="sett"
            patternUnits="userSpaceOnUse"
            width="1000"
            x="0"
            y="0"
          >
            <g
              id="weft"
            >
              <rect
                fill="#006818"
                height="518.5185185185185"
                width="100%"
                x="0"
                y="0"
              />
              <rect
                fill="#FCFCFC"
                height="111.11111111111111"
                width="100%"
                x="0"
                y="518.5185185185185"
              />
              <rect
                fill="#202060"
                height="259.25925925925924"
                width="100%"
                x="0"
                y="629.6296296296296"
              />
              <rect
                fill="#FCFCFC"
                height="111.11111111111111"
                width="100%"
                x="0"
                y="888.8888888888888"
              />
            </g>
            <g
              id="warp"
              mask="url(#threadHatchingMask)"
            >
              <rect
                fill="#006818"
                height="100%"
                width="518.5185185185185"
                x="0"
                y="0"
              />
              <rect
                fill="#FCFCFC"
                height="100%"
                width="111.11111111111111"
                x="518.5185185185185"
                y="0"
              />
              <rect
                fill="#202060"
                height="100%"
                width="259.25925925925924"
                x="629.6296296296296"
                y="0"
              />
              <rect
                fill="#FCFCFC"
                height="100%"
                width="111.11111111111111"
                x="888.8888888888888"
                y="0"
              />
            </g>
          </pattern>
        </defs>
        <rect
          data-testid="tartanRect"
          fill="url(#sett)"
          filter=""
          height="100%"
          width="100%"
          x="0"
          y="0"
        />
      </svg>
    </div>
  </body>,
  "container": <div>
    <svg
      data-testid="tartanImage"
      height="1000"
      id="testId"
      style="background-color: red;"
      width="1000"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <mask
          height="1"
          id="threadHatchingMask"
          width="1"
          x="0"
          y="0"
        >
          <rect
            fill="url(#threadHatching)"
            height="100%"
            width="100%"
            x="0"
            y="0"
          />
        </mask>
        <pattern
          height="18.51851851851852"
          id="threadHatching"
          patternUnits="userSpaceOnUse"
          width="18.51851851851852"
          x="0"
          y="0"
        >
          <rect
            fill="white"
            height="9.25925925925926"
            width="4.62962962962963"
            x="0"
            y="-4.62962962962963"
          />
          <rect
            fill="white"
            height="9.25925925925926"
            width="4.62962962962963"
            x="0"
            y="13.88888888888889"
          />
          <rect
            fill="white"
            height="9.25925925925926"
            width="4.62962962962963"
            x="4.62962962962963"
            y="0"
          />
          <rect
            fill="white"
            height="9.25925925925926"
            width="4.62962962962963"
            x="9.25925925925926"
            y="4.62962962962963"
          />
          <rect
            fill="white"
            height="9.25925925925926"
            width="4.62962962962963"
            x="13.88888888888889"
            y="9.25925925925926"
          />
        </pattern>
        <pattern
          data-testid="testSett"
          height="1000"
          id="sett"
          patternUnits="userSpaceOnUse"
          width="1000"
          x="0"
          y="0"
        >
          <g
            id="weft"
          >
            <rect
              fill="#006818"
              height="518.5185185185185"
              width="100%"
              x="0"
              y="0"
            />
            <rect
              fill="#FCFCFC"
              height="111.11111111111111"
              width="100%"
              x="0"
              y="518.5185185185185"
            />
            <rect
              fill="#202060"
              height="259.25925925925924"
              width="100%"
              x="0"
              y="629.6296296296296"
            />
            <rect
              fill="#FCFCFC"
              height="111.11111111111111"
              width="100%"
              x="0"
              y="888.8888888888888"
            />
          </g>
          <g
            id="warp"
            mask="url(#threadHatchingMask)"
          >
            <rect
              fill="#006818"
              height="100%"
              width="518.5185185185185"
              x="0"
              y="0"
            />
            <rect
              fill="#FCFCFC"
              height="100%"
              width="111.11111111111111"
              x="518.5185185185185"
              y="0"
            />
            <rect
              fill="#202060"
              height="100%"
              width="259.25925925925924"
              x="629.6296296296296"
              y="0"
            />
            <rect
              fill="#FCFCFC"
              height="100%"
              width="111.11111111111111"
              x="888.8888888888888"
              y="0"
            />
          </g>
        </pattern>
      </defs>
      <rect
        data-testid="tartanRect"
        fill="url(#sett)"
        filter=""
        height="100%"
        width="100%"
        x="0"
        y="0"
      />
    </svg>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <svg
        data-testid="tartanImage"
        height="1000"
        id="testId"
        style="background-color: red;"
        width="1000"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <mask
            height="1"
            id="threadHatchingMask"
            width="1"
            x="0"
            y="0"
          >
            <rect
              fill="url(#threadHatching)"
              height="100%"
              width="100%"
              x="0"
              y="0"
            />
          </mask>
          <pattern
            height="18.51851851851852"
            id="threadHatching"
            patternUnits="userSpaceOnUse"
            width="18.51851851851852"
            x="0"
            y="0"
          >
            <rect
              fill="white"
              height="9.25925925925926"
              width="4.62962962962963"
              x="0"
              y="-4.62962962962963"
            />
            <rect
              fill="white"
              height="9.25925925925926"
              width="4.62962962962963"
              x="0"
              y="13.88888888888889"
            />
            <rect
              fill="white"
              height="9.25925925925926"
              width="4.62962962962963"
              x="4.62962962962963"
              y="0"
            />
            <rect
              fill="white"
              height="9.25925925925926"
              width="4.62962962962963"
              x="9.25925925925926"
              y="4.62962962962963"
            />
            <rect
              fill="white"
              height="9.25925925925926"
              width="4.62962962962963"
              x="13.88888888888889"
              y="9.25925925925926"
            />
          </pattern>
          <pattern
            data-testid="testSett"
            height="1000"
            id="sett"
            patternUnits="userSpaceOnUse"
            width="1000"
            x="0"
            y="0"
          >
            <g
              id="weft"
            >
              <rect
                fill="#006818"
                height="518.5185185185185"
                width="100%"
                x="0"
                y="0"
              />
              <rect
                fill="#FCFCFC"
                height="111.11111111111111"
                width="100%"
                x="0"
                y="518.5185185185185"
              />
              <rect
                fill="#202060"
                height="259.25925925925924"
                width="100%"
                x="0"
                y="629.6296296296296"
              />
              <rect
                fill="#FCFCFC"
                height="111.11111111111111"
                width="100%"
                x="0"
                y="888.8888888888888"
              />
            </g>
            <g
              id="warp"
              mask="url(#threadHatchingMask)"
            >
              <rect
                fill="#006818"
                height="100%"
                width="518.5185185185185"
                x="0"
                y="0"
              />
              <rect
                fill="#FCFCFC"
                height="100%"
                width="111.11111111111111"
                x="518.5185185185185"
                y="0"
              />
              <rect
                fill="#202060"
                height="100%"
                width="259.25925925925924"
                x="629.6296296296296"
                y="0"
              />
              <rect
                fill="#FCFCFC"
                height="100%"
                width="111.11111111111111"
                x="888.8888888888888"
                y="0"
              />
            </g>
          </pattern>
        </defs>
        <rect
          data-testid="tartanRect"
          fill="url(#sett)"
          filter=""
          height="100%"
          width="100%"
          x="0"
          y="0"
        />
      </svg>
    </div>
  </body>,
  "container": <div>
    <svg
      data-testid="tartanImage"
      height="1000"
      id="testId"
      style="background-color: red;"
      width="1000"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <mask
          height="1"
          id="threadHatchingMask"
          width="1"
          x="0"
          y="0"
        >
          <rect
            fill="url(#threadHatching)"
            height="100%"
            width="100%"
            x="0"
            y="0"
          />
        </mask>
        <pattern
          height="18.51851851851852"
          id="threadHatching"
          patternUnits="userSpaceOnUse"
          width="18.51851851851852"
          x="0"
          y="0"
        >
          <rect
            fill="white"
            height="9.25925925925926"
            width="4.62962962962963"
            x="0"
            y="-4.62962962962963"
          />
          <rect
            fill="white"
            height="9.25925925925926"
            width="4.62962962962963"
            x="0"
            y="13.88888888888889"
          />
          <rect
            fill="white"
            height="9.25925925925926"
            width="4.62962962962963"
            x="4.62962962962963"
            y="0"
          />
          <rect
            fill="white"
            height="9.25925925925926"
            width="4.62962962962963"
            x="9.25925925925926"
            y="4.62962962962963"
          />
          <rect
            fill="white"
            height="9.25925925925926"
            width="4.62962962962963"
            x="13.88888888888889"
            y="9.25925925925926"
          />
        </pattern>
        <pattern
          data-testid="testSett"
          height="1000"
          id="sett"
          patternUnits="userSpaceOnUse"
          width="1000"
          x="0"
          y="0"
        >
          <g
            id="weft"
          >
            <rect
              fill="#006818"
              height="518.5185185185185"
              width="100%"
              x="0"
              y="0"
            />
            <rect
              fill="#FCFCFC"
              height="111.11111111111111"
              width="100%"
              x="0"
              y="518.5185185185185"
            />
            <rect
              fill="#202060"
              height="259.25925925925924"
              width="100%"
              x="0"
              y="629.6296296296296"
            />
            <rect
              fill="#FCFCFC"
              height="111.11111111111111"
              width="100%"
              x="0"
              y="888.8888888888888"
            />
          </g>
          <g
            id="warp"
            mask="url(#threadHatchingMask)"
          >
            <rect
              fill="#006818"
              height="100%"
              width="518.5185185185185"
              x="0"
              y="0"
            />
            <rect
              fill="#FCFCFC"
              height="100%"
              width="111.11111111111111"
              x="518.5185185185185"
              y="0"
            />
            <rect
              fill="#202060"
              height="100%"
              width="259.25925925925924"
              x="629.6296296296296"
              y="0"
            />
            <rect
              fill="#FCFCFC"
              height="100%"
              width="111.11111111111111"
              x="888.8888888888888"
              y="0"
            />
          </g>
        </pattern>
      </defs>
      <rect
        data-testid="tartanRect"
        fill="url(#sett)"
        filter=""
        height="100%"
        width="100%"
        x="0"
        y="0"
      />
    </svg>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
